crates/red_knot_python_semantic/src/types/mro.rs

```diff
@@ -0,0 +1,641 @@
+                            .collect(),
+                    );
+                    match linearized {
+                        Some(mro) => {
+                            mro_possibilities.insert(mro);
+                        },
+                        None => {
+                            if !mro_errors.contains(&bases_possibility) {
+                                mro_errors.insert(bases_possibility.clone());
+                            }
+                        },
+                    };
+                }
+            }
+        }
+    }
+
+    MroPossibilities::possibly_many(mro_possibilities, mro_errors)
+}
+
+/// Given a list of types representing the bases of a class,
+/// of which one or more types could be a [`Type::Union`] variant,
+/// resolve the list into a "union of bases lists", where each list in the union
+/// is guaranteed not to hold any bases that are a [`Type::Union`].
+fn fork_bases<'db>(db: &'db dyn Db, bases: &[Type<'db>]) -> BasesPossibilities<'db> {
+    // Fast path for the common case, where none of the bases is a [`Type::Union`]:
+    if !bases.iter().any(Type::is_union) {
+        return BasesPossibilities::Single(bases.iter().map(ClassBase::from).collect());
+    }
+
+    // Slow path: one or more of the bases is a [`Type::Union`]
+    let mut possibilities = FxHashSet::from_iter([BasesList::default()]);
+    for base in bases {
+        possibilities = add_next_base(db, &possibilities, *base);
+    }
+    BasesPossibilities::Multiple(possibilities)
+}
+
+fn add_next_base<'db>(
+    db: &'db dyn Db,
+    bases_possibilities: &FxHashSet<BasesList<'db>>,
+    next_base: Type<'db>,
+) -> FxHashSet<BasesList<'db>> {
+    let mut new_possibilities = FxHashSet::default();
+    let mut add_non_union_base = |fork: &[ClassBase<'db>], base: Type<'db>| {
+        new_possibilities.insert(
+            fork.iter()
+                .copied()
+                .chain(std::iter::once(ClassBase::from(base)))
+                .collect(),
+        );
+    };
+    match next_base {
+        Type::Union(union) => {
+            for element in union.elements(db) {
+                for existing_possibility in bases_possibilities {
+                    add_non_union_base(existing_possibility, *element);
+                }
+            }
+        }
+        _ => {
+            for possibility in bases_possibilities {
+                add_non_union_base(possibility, next_base);
+            }
+        }
+    }
+    debug_assert_ne!(new_possibilities.len(), 0);
+    new_possibilities
+}
+
+/// The possible value of `__bases__` for a given class.
+///
+/// Whereas [`ClassType::bases`] returns a list of types in which any type
+/// might be a [`Type::Union`], this enum transforms the list of types so that we
+/// have a union of possible `__bases__` lists rather than a single list
+/// that could contain a union.
+enum BasesPossibilities<'db> {
```

Comment:
I'm struggling a bit with the names. There are so many `Bases` and lists. I think a module-level comment explaining the different types and how this all fits together would be useful.